From 79223682fdfebab33a2dbe2f2c51eb23d99d22f8 Mon Sep 17 00:00:00 2001
From: "Dr. Greg" <greg@enjellic.com>
Date: Fri, 30 Dec 2016 11:35:13 -0600
Subject: [PATCH 01/24] Remove IMA status on inode unlink.

This update integrates modifies the LSM hook which is invoked on                inode removal to call into the integrity inode processing logic                 to clear the IMA flag on the inode when it is removed.                                                                                                          This is to support the notion of file pseudonyms which have their               measurement value populated as a component of the system behavior               setup.  Clearing the IMA attributes on the inode will cause a new               measurement value to be created for the file if the inode is                    unlinked after the pseudonym definition has been set on the file.
---
 security/integrity/iint.c | 5 +++++
 security/security.c       | 1 +
 2 files changed, 6 insertions(+)

diff --git a/security/integrity/iint.c b/security/integrity/iint.c
index 3d2f5b45c8cb..3aaac2b425ce 100644
--- a/security/integrity/iint.c
+++ b/security/integrity/iint.c
@@ -138,13 +138,18 @@ void integrity_inode_free(struct inode *inode)
 
 	if (!IS_IMA(inode))
 		return;
+	pr_info("[%s]: Processing eviction for inode=%ld, flags=%u\n",
+		__func__, inode->i_ino, inode->i_flags);
 
 	write_lock(&integrity_iint_lock);
 	iint = __integrity_iint_find(inode);
 	rb_erase(&iint->rb_node, &integrity_iint_tree);
 	write_unlock(&integrity_iint_lock);
 
+	pr_info("[%s]: Evicting security inode %p for %ld.\n", __func__,
+		iint, iint->inode->i_ino);
 	iint_free(iint);
+	inode->i_flags &= ~S_IMA;
 }
 
 static void init_once(void *foo)
diff --git a/security/security.c b/security/security.c
index 46f405ce6b0f..d06466fda29a 100644
--- a/security/security.c
+++ b/security/security.c
@@ -525,6 +525,7 @@ int security_inode_unlink(struct inode *dir, struct dentry *dentry)
 {
 	if (unlikely(IS_PRIVATE(d_backing_inode(dentry))))
 		return 0;
+	integrity_inode_free(dentry->d_inode);
 	return call_int_hook(inode_unlink, 0, dir, dentry);
 }
 
-- 
2.11.0

